pkg/querier/tail.go

```diff
@@ -103,22 +104,32 @@ func (t *Tailer) readTailClients() {
 // keeps sending oldest entry to responseChan. If channel is blocked drop the entry
 // When channel is unblocked, send details of dropped entries with current entry
 func (t *Tailer) loop() {
-	ticker := time.NewTicker(checkConnectionsWithIngestersPeriod)
-	defer ticker.Stop()
+	checkConnectionTicker := time.NewTicker(checkConnectionsWithIngestersPeriod)
+	defer checkConnectionTicker.Stop()
+
+	tailLengthLimitTicker := time.NewTicker(t.tailLengthLimit)
+	defer tailLengthLimitTicker.Stop()
 
 	tailResponse := new(TailResponse)
 
+outer:
 	for {
 		if t.stopped {
 			break
 		}
 
 		select {
-		case <-ticker.C:
+		case <-checkConnectionTicker.C:
 			// Try to reconnect dropped ingesters and connect to new ingesters
 			if err := t.checkIngesterConnections(); err != nil {
 				level.Error(util.Logger).Log("Error reconnecting to disconnected ingesters", fmt.Sprintf("%v", err))
 			}
+		case <-tailLengthLimitTicker.C:
+			if err := t.close(); err != nil {
+				level.Error(util.Logger).Log("Error closing Tailer", fmt.Sprintf("%v", err))
+			}
+			t.closeErrChan <- errors.New("Reached tail length limit")
+			break outer
```

Comment:
would just a `return` here be simpler?

Reply:
Consumer of tail request would not know about why tail request ended. That's why I did this.

Reply:
I'm referring to the `break outer` would a `return` here work but be a little simpler?

Reply:
Ohh sorry, my bad. Actually I did not like mix of return and break statement. Everywhere else there is a break. Should change everything to return? What do you think?

Reply:
return is less prone to error in my opinion, if they all can return immediately of course.